src/mem_backed_tree.rs

```diff
@@ -124,33 +116,49 @@ where
         }
 
         // Recalculate the tree starting at the new leaf
-        let num_leaves = self.leaves.len();
+        let num_leaves = self.len();
         let new_leaf_idx = LeafIdx::new(num_leaves - 1);
+        // recalculate_path() requires its leaf idx to be less than usize::MAX. This is guaranteed
+        // because it's self.len() - 1.
         self.recaluclate_path(new_leaf_idx)
     }
 
     /// Checks that this tree is well-formed. This can take a while if the tree is large. Run this
     /// if you've deserialized this tree and don't trust the source. If a tree is malformed, other
     /// methods will panic or behave oddly.
     pub fn self_check(&self) -> Result<(), SelfCheckError> {
-        // Go through every level of the tree, checking hashes
-        let num_leaves = self.leaves.len();
+        // If the number of leaves is more than an in-memory tree could support, return an error
+        let num_leaves = self.len();
+        if num_leaves > (usize::MAX / 2) as u64 {
+            return Err(SelfCheckError::TooManyLeaves);
+        }
+
+        // If the number of internal nodes is less than the necessary size of the tree, return an error
         let num_nodes = num_internal_nodes(num_leaves);
+        if (self.internal_nodes.len() as u64) < num_nodes {
+            return Err(SelfCheckError::MissingNode(self.internal_nodes.len() as u64));
+        }
+        // If the number of internal nodes exceeds the necessary size of the tree, return an error
+        if (self.internal_nodes.len() as u64) > num_nodes {
+            return Err(SelfCheckError::TooManyInternalNodes);
+        }
 
         // Start on level 0. We check the leaf hashes
         for (leaf_idx, leaf) in self.leaves.iter().enumerate() {
-            let leaf_hash_idx: InternalIdx = LeafIdx::new(leaf_idx).into();
+            let leaf_hash_idx: InternalIdx = LeafIdx::new(leaf_idx as u64).into();
 
             // Compute the leaf hash and retrieve the stored leaf hash
             let expected_hash = leaf_hash::<H, _>(leaf);
-            let stored_hash = match self.internal_nodes.get(leaf_hash_idx.as_usize()) {
-                None => Err(SelfCheckError::MissingNode(leaf_hash_idx.as_usize())),
+            // We can unwrap() because we checked above that the number of nodes necessary for this
+            // tree fits in memory
+            let stored_hash = match self.internal_nodes.get(leaf_hash_idx.as_usize().unwrap()) {
```

Comment:
`clippy::pedantic` would probably suggest to rewrite this using `let..else` to something like:
```
let Some(stored_hash) = self.internal_nodes.get(leaf_hash_idx.as_usize().unwrap()) else {
  return Err(SelfCheckError::MissingNode(leaf_hash_idx.as_u64()));
};
```